pkg/util/env/env.go

```diff
@@ -0,0 +1,138 @@
+func NewResourceStore() *ResourceStore {
+	return &ResourceStore{
+		SecretStore:    make(map[string]*kapi.Secret),
+		ConfigMapStore: make(map[string]*kapi.ConfigMap),
+	}
+}
+
+// getSecretRefValue returns the value of a secret in the supplied namespace
+func getSecretRefValue(client kclientset.Interface, namespace string, store *ResourceStore, secretSelector *kapi.SecretKeySelector) (string, error) {
+	secret, ok := store.SecretStore[secretSelector.Name]
+	if !ok {
+		var err error
+		secret, err = client.Core().Secrets(namespace).Get(secretSelector.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		store.SecretStore[secretSelector.Name] = secret
+	}
+	if data, ok := secret.Data[secretSelector.Key]; ok {
+		return string(data), nil
+	}
+	return "", fmt.Errorf("key %s not found in secret %s", secretSelector.Key, secretSelector.Name)
+
+}
+
+// getConfigMapRefValue returns the value of a configmap in the supplied namespace
+func getConfigMapRefValue(client kclientset.Interface, namespace string, store *ResourceStore, configMapSelector *kapi.ConfigMapKeySelector) (string, error) {
+	configMap, ok := store.ConfigMapStore[configMapSelector.Name]
+	if !ok {
+		var err error
+		configMap, err = client.Core().ConfigMaps(namespace).Get(configMapSelector.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		store.ConfigMapStore[configMapSelector.Name] = configMap
+	}
+	if data, ok := configMap.Data[configMapSelector.Key]; ok {
+		return string(data), nil
+	}
+	return "", fmt.Errorf("key %s not found in config map %s", configMapSelector.Key, configMapSelector.Name)
+}
+
+// getFieldRef returns the value of the supplied path in the given object
+func getFieldRef(obj runtime.Object, from *kapi.EnvVarSource) (string, error) {
+	return fieldpath.ExtractFieldPathAsString(obj, from.FieldRef.FieldPath)
+}
+
+// getResourceFieldRef returns the value of a resource in the given container
+func getResourceFieldRef(from *kapi.EnvVarSource, c *kapi.Container) (string, error) {
+	return fieldpath.InternalExtractContainerResourceValue(from.ResourceFieldRef, c)
+}
+
+// GenEnvVarRefValue returns the value referenced by the supplied EnvVarSource given the other supplied information
+func GetEnvVarRefValue(f *clientcmd.Factory, kc kclientset.Interface, ns string, store *ResourceStore, from *kapi.EnvVarSource, obj runtime.Object, c *kapi.Container) (string, error) {
+	var kubeClient kclientset.Interface
+	var namespace string
+	var err error
+
+	if f != nil {
+		kubeClient, err = f.ClientSet()
+		if err != nil {
+			return "", err
+		}
+		namespace, _, err = f.DefaultNamespace()
+		if err != nil {
+			return "", err
+		}
+	} else if kc != nil {
+		kubeClient = kc
+		namespace = ns
+	} else {
+		return "", fmt.Errorf("You must supply either a clientcmd.Factory or kclientset.Interface")
+	}
+
+	if from.SecretKeyRef != nil {
+		return getSecretRefValue(kubeClient, namespace, store, from.SecretKeyRef)
```

Comment:
This method is a security escalation attack waiting to happen.  Please add specific separate tests that verify that no call outside of `ns` is made, and add specific language to the call sites that doing so is not allowed.